src/ameba.ts

```diff
@@ -27,23 +28,68 @@ export class Ameba {
         this.config = getConfig();
     }
 
-    public execute(document: TextDocument): void {
+    public execute(document: TextDocument, virtual = false): void {
         if (document.languageId !== 'crystal' || document.isUntitled || document.uri.scheme !== 'file') {
             return;
         }
 
-        const args = [this.config.command, document.fileName, '--format', 'json'];
-        const dir = workspace.getWorkspaceFolder(document.uri)!.uri.fsPath;
-        const configFile = path.join(dir, this.config.configFileName);
+        const space = workspace.getWorkspaceFolder(document.uri) ?? noWorkspaceFolder(document);
+        const args = [this.config.command];
+
+        // When reading from stdin, cannot pass any files via the CLI, otherwise there will be overlaps
+        // and potentially out of date errors
+        if (!virtual) {
+            args.push(document.fileName);
+        } else {
+            args.push('--stdin-filename', document.uri.fsPath);
```

Comment:
I'm wondering whether passing this flag is actually _required_?

Reply:
I guess technically it may not be if we assume the ameba response is for our document. It is necessary for ameba though as that's how we tell it to read from STDIN.

Reply:
Since you're sending one document at a time this shouldn't be needed really.
By removing this you can also drop the `semver` dependency and all of the code around it, that's restricting the `onType` support to ameba >= 1.6.2.

Reply:
I'm not sure I follow. We need this flag in order to tell ameba to read from STDIN, which is necessary for linting while typing to work.

Reply:
No, ameba accepts stdin regardless of this flag. This is just for _naming_ the output source.

Reply:
Hmm, now I'm not sure myself, although I think that shouldn't be needed, as you can pass `-` in order to read from STDIN anyway.

Reply:
```sh
→ ameba --version                                                                                                                               
1.6.1
```

```sh
→ cat spec/ameba/rule/lint/typos_spec.cr | ameba --format json -                                                                                       
{"sources":[],"metadata":{"ameba_version":"1.6.1","crystal_version":"1.11.0"},"summary":{"target_sources_count":0,"issues_count":0}}
```

Reply:
Hmm, nope... it doesn't work like this, disregard my comments here. Need to investigate further.

Reply:
[This](https://github.com/crystal-ameba/ameba/blob/a5c99e2359bcebc70a3800353fb92036a3835f81/src/ameba/config.cr#L171) is the only place in the source code ameba reads from STDIN.

That example is just the default ameba output when no sources are passed. `sources` is empty.

```sh
$ ameba --format json -                                                                                       
{"sources":[],"metadata":{"ameba_version":"1.6.1","crystal_version":"1.11.0"},"summary":{"target_sources_count":0,"issues_count":0}}

$ cat spec/ameba/rule/lint/typos_spec.cr | ameba --format json --stdin-filename typos_spec.cr /dev/null
{"sources":[{"path":"typos_spec.cr","issues":[]}],"metadata":{"ameba_version":"1.6.3","crystal_version":"1.14.0"},"summary":{"target_sources_count":1,"issues_count":0}}
```

Reply:
Damn, there's a bug in the implementation causing the flag to _add_ the passed STDIN file to the sources instead of just linting just this one... I'm gonna fix it.

Reply:
https://github.com/crystal-ameba/ameba/pull/501

Reply:
It shouldn't exclusively use the file passed in from STDIN as users can pass other files in from the CLI to lint alongside it. Relevant commit: https://github.com/crystal-ameba/ameba/commit/f29cffa83b7efc3437419ecac9a6a5c83af4e39a

The "bug" is that when not passing in any files via CLI, it searches for ones in the current directory.

Reply:
That's IMO flawed logic. There's no point in combining these two. You either pass sth through the `STDIN` or as a filename list.

Reply:
That's fair. I just didn't want to arbitrarily limit what could be done w/ ameba.

> That's exactly what you're doing in this PR btw.
I know